frontend/server/app.ts

```diff
@@ -0,0 +1,161 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import * as path from 'path';
+import * as express from 'express';
+import { Application, static as StaticHandler } from 'express';
+import * as proxy from 'http-proxy-middleware';
+
+import { IConfigs } from './configs';
+import { getAddress } from './utils';
+import { getBuildMetadata, getHealthzEndpoint, getHealthzHandler } from './handlers/healthz';
+import { getArtifactsHandler } from './handlers/artifacts';
+import { getCreateTensorboardHandler, getGetTensorboardHandler } from './handlers/tensorboard';
+import { getPodLogsHandler } from './handlers/pod-logs';
+import { clusterNameHandler, projectIdHandler } from './handlers/core';
+import { getAllowCustomVisualizationsHandler } from './handlers/vis';
+import { getIndexHTMLHandler } from './handlers/index-html';
+
+import proxyMiddleware from './proxy-middleware';
+import { Server } from 'http';
+
+function getRegisterHandler(app: Application, basePath: string) {
+  return (
+    func: (name: string, handler: express.Handler) => express.Application,
+    route: string,
+    handler: express.Handler,
+  ) => {
+    func.call(app, route, handler);
+    return func.call(app, `${basePath}${route}`, handler);
+  };
+}
+
+export class UIServer {
+  app: Application;
+  httpServer?: Server;
+
+  constructor(public readonly options: IConfigs) {
```

Comment:
I'm not familiar with node common patterns, is using constructor with heavy initialization logic common?
I'd personally prefer using a pure function, e.g. `initUIServer(options)` to create an instance of UIServer. Extra benefit is that, the creator function can be async too (I guess it's arguable whether we need this or not).

This is just discussion. What do you think?

Reply:
classes are just syntactic sugars in js. I think there are no right common patterns (as in it is fine either way). 

The main reason I did this is so that I write less code in my unit test. It is easier for me to do the setup and teardown if I only need to track a single object.

But I think this is perfectly fine too.
```js
const app = initUIServer(options);
const server = app.listen(port, ()=>console.info(`listening to port:${port}.`);
server.close();
```